core/common/src/main/java/alluxio/network/connection/ThriftClientPool.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.network.connection;
+
+import alluxio.Constants;
+import alluxio.RuntimeConstants;
+import alluxio.exception.ExceptionMessage;
+import alluxio.resource.DynamicResourcePool;
+import alluxio.retry.ExponentialBackoffRetry;
+import alluxio.retry.RetryPolicy;
+import alluxio.security.authentication.TransportProvider;
+import alluxio.thrift.AlluxioService;
+
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TBinaryProtocol;
+import org.apache.thrift.protocol.TMultiplexedProtocol;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.TTransport;
+import org.apache.thrift.transport.TTransportException;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.regex.Pattern;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ *  A pool to manage Alluxio thrift clients.
+ *
+ * @param <T> the Alluxio thrift service type
+ */
+@ThreadSafe
+public abstract class ThriftClientPool<T extends AlluxioService.Client>
```

Comment:
The logic is mostly tested in DynamicResourcePoolTest.java

This class is pretty long but actually doesn't have much logic. All it does it to open the connection and close the connection. In this sense, it is not critical to have a unit test. This class is excised in the integration test.  I have added a TODO to add unittest which is good to have as you said.  I will address this TODO after 1.3 release.  Let me know if you want the unittest to be included in this release, I can think about that and do it as well